src/batch.jl

```diff
@@ -0,0 +1,199 @@
+            xbuf = similar(xtmp, typeof(one(eltype(f.x))))
+            ybuf = similar(ytmp, typeof(oneunit(eltype(f.y))*oneunit(s1)))
+            if inf1 && inf2 # x = t/(1-t^2) coordinate transformation
+                return workfunc(BatchIntegrand((v, t) -> begin resize!(xtmp, length(t)); resize!(ytmp, length(v));
+                                            f.f!(ytmp, xtmp .= oneunit(s1) .* t ./ (1 .- t .* t)); v .= ytmp .* (1 .+ t .* t) .* oneunit(s1) ./ (1 .- t .* t) .^ 2; end, ybuf, xbuf, f.max_batch),
+                                map(x -> isinf(x) ? (signbit(x) ? -one(x) : one(x)) : 2x / (oneunit(x)+hypot(oneunit(x),2x)), s),
+                                t -> oneunit(s1) * t / (1 - t^2))
+            end
+            let (s0,si) = inf1 ? (s2,s1) : (s1,s2) # let is needed for JuliaLang/julia#15276
+                if si < zero(si) # x = s0 - t/(1-t)
+                    return workfunc(BatchIntegrand((v, t) -> begin resize!(xtmp, length(t)); resize!(ytmp, length(v));
+                                            f.f!(ytmp, xtmp .= s0 .- oneunit(s1) .* t ./ (1 .- t)); v .= ytmp .* oneunit(s1) ./ (1 .- t) .^ 2; end, ybuf, xbuf, f.max_batch),
+                                    reverse(map(x -> 1 / (1 + oneunit(x) / (s0 - x)), s)),
+                                    t -> s0 - oneunit(s1)*t/(1-t))
+                else # x = s0 + t/(1-t)
+                    return workfunc(BatchIntegrand((v, t) -> begin resize!(xtmp, length(t)); resize!(ytmp, length(v));
+                                            f.f!(ytmp, xtmp .= s0 .+ oneunit(s1) .* t ./ (1 .- t)); v .= ytmp .* oneunit(s1) ./ (1 .- t) .^ 2; end, ybuf, xbuf, f.max_batch),
+                                    map(x -> 1 / (1 + oneunit(x) / (x - s0)), s),
+                                    t -> s0 + oneunit(s1)*t/(1-t))
+                end
+            end
+        end
+    end
+    return workfunc(f, s, identity)
+end
+
+"""
+    quadgk(f::BatchIntegrand, a,b,c...; kws...)
+
+Like [`quadgk`](@ref), but batches evaluation points for an in-place integrand to evaluate
+simultaneously. In particular, there are two differences from `quadgk`
+
+1. The function `f.f!` should be of the form `f!(y, x) = y .= f.(x)`.  That is, it writes
+   the return values of the integand `f(x)` in-place into its first argument `y`. (The
+   return value of `f!` is ignored.) See [`BatchIntegrand`](@ref) for how to define the
+   integrand.
+
+2. `f.max_batch` changes how the adaptive refinement is done by batching multiple segments
+   together. Choosing `max_batch<=4*order+2` will reproduce the result of
+   non-batched `quadgk`, however if `max_batch=n*(4*order+2)` up to `2n` Kronrod rules will be evaluated
+   together, which can produce slightly different results and sometimes require more
+   integrand evaluations when using relative tolerances.
+"""
+function quadgk(f::BatchIntegrand{F,Y,<:AbstractVector{Nothing}}, segs::T...; kws...) where {F,Y,T}
```

Comment:
Should this be a method of `quadgk!` rather than `quadgk`, since it modifies the buffers in `f`?

Reply:
On the other hand, it doesn't have a pre-allocated `result` argument.